streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreTest.java

```diff
@@ -0,0 +1,435 @@
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 1, "vn1", SEGMENT_INTERVAL - 1);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 2, "vn2", SEGMENT_INTERVAL - 2);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 5, "vn10", SEGMENT_INTERVAL - 10);
+    }
+
+    @Test
+    public void shouldMoveRecordToOlderSegmentWithNullsDuringPut() {
+        putToStore("k", null, SEGMENT_INTERVAL + 20);
+        putToStore("k", null, SEGMENT_INTERVAL - 1);
+        putToStore("k", null, SEGMENT_INTERVAL + 1);
+        putToStore("k", null, SEGMENT_INTERVAL - 10);
+        putToStore("k", null, SEGMENT_INTERVAL + 10);
+        putToStore("k", "vp5", SEGMENT_INTERVAL + 5);
+        putToStore("k", "vn5", SEGMENT_INTERVAL - 5);
+        putToStore("k", "vn6", SEGMENT_INTERVAL - 6);
+
+        verifyGetNullFromStore("k");
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL + 30);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL + 15);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL + 6, "vp5", SEGMENT_INTERVAL + 5);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL + 2);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 1);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 5, "vn5", SEGMENT_INTERVAL - 5);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 6, "vn6", SEGMENT_INTERVAL - 6);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 8);
+    }
+
+    @Test
+    public void shouldFallThroughToExistingOlderSegmentAsLatestDuringPut() {
+        putToStore("k", null, SEGMENT_INTERVAL - 5);
+        putToStore("k", "vn6", SEGMENT_INTERVAL - 6);
+        putToStore("k", "vp20", SEGMENT_INTERVAL + 20);
+        putToStore("k", null, SEGMENT_INTERVAL + 10);
+        putToStore("k", null, SEGMENT_INTERVAL - 1);
+        putToStore("k", "vn2", SEGMENT_INTERVAL - 2);
+
+        verifyGetValueFromStore("k", "vp20", SEGMENT_INTERVAL + 20);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL + 30, "vp20", SEGMENT_INTERVAL + 20);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL + 12);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 1);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 2, "vn2", SEGMENT_INTERVAL - 2);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 5);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 6, "vn6", SEGMENT_INTERVAL - 6);
+    }
+
+    @Test
+    public void shouldPutNonLatestTombstoneIntoNewSegmentWithValidTo() {
+        putToStore("k", "vp30", SEGMENT_INTERVAL + 30);
+        putToStore("k", null, SEGMENT_INTERVAL - 10); // this put should result in tombstone with validTo=SEGMENT_INTERVAL+30
```

Comment:
This is the degenerated segment case, right? So `validTo` should be `validfrom == SEGMENT_INTERVAL - 10`? (or do I remember incorrectly how degenerated segments work?